src/gerbil/runtime/mop.ss

```diff
@@ -229,174 +242,113 @@ namespace: #f
-          (else (lp rest ctor))))
-        (else ctor))))
-
-  (def (find-super-struct super)
-    (def (base-struct super-struct klass)
-      (cond
-       (super-struct
-        (cond
-         ((struct-subtype? super-struct klass)
-          (let lp ((klass klass))
-            (if (struct-type? klass)
-              klass
-              (lp (##type-super klass)))))
-         ((struct-subtype? klass super-struct)
-          super-struct)
-         (else
-          (error "Bad mixin: incompatible struct bases" klass super-struct))))
-       ((struct-type? klass) klass)
-       ((class-type? klass)
-        (let lp ((next (##type-super klass)))
-          (cond
-           ((not next)
-            #f)
-           ((struct-type? next)
-            next)
-           ((class-type? next)
-            (lp next))
-           (else #f))))
-       (else #f)))
-
-    (let lp ((rest super) (super-struct #f))
-      (match rest
-        ([hd . rest]
-         (lp rest (base-struct super-struct hd)))
-        (else super-struct))))
-
-  (def (expand-struct-mixin super)
-    (let lp ((rest super) (mixin '()))
-      (match rest
-        ([hd . rest]
-         (if (struct-type? hd)
-           (let lp2 ((next hd) (mixin mixin))
-             (cond
-              ((not next)
-               (lp rest mixin))
-              ((struct-type? next)
-               (lp2 (##type-super next) (cons next mixin)))
-              (else
-               (lp rest mixin))))
-           (lp rest (cons hd mixin))))
-        (else
-         (reverse mixin)))))
-
   (cond
    ((find (lambda (klass) (not (type-descriptor? klass))) super)
-    => (lambda (klass)
-         (error "Illegal super class; not a type descriptor" klass)))
-   ((find (lambda (klass)
-             (assgetq final: (type-descriptor-plist klass)))
-           super)
-    => (lambda (klass)
-         (error "Cannot extend final class" klass))))
-
-  (let* ((std-super (find-super-struct super))
-         (mixin (if std-super (expand-struct-mixin super) super)))
-    (let-values (((std-fields std-slots std-slot-list)
-                  (make-slots (if std-super (type-descriptor-fields std-super) 0))))
-      (let* ((std-mixin  (class-linearize-mixins mixin))
-             (std-plist  (if std-super
-                           (let (fields (assgetq fields: (type-descriptor-plist std-super)))
-                             (cons (cons fields: fields) plist))
-                           plist))
-             (std-plist  (cons (cons slots: std-slot-list) std-plist))
-             (std-ctor   (or ctor (find-super-ctor super))))
-        (make-class-type-descriptor id name std-super std-mixin std-fields std-plist std-ctor std-slots)))))
+    => (cut error "Illegal super class; not a type descriptor" <>))
+   ((find (lambda (klass) (assgetq final: (type-descriptor-plist klass))) super)
+    => (cut error "Cannot extend final class" <>)))
+
+  (let*-values (((std-super) (apply base-struct super)) ;; super struct, if any
```

Comment:
You can just `let*` and `((values ...) foo)`.

Reply:
Wow, I didn't know that. Very cool!